prboom2/src/d_main.c

```diff
@@ -1527,9 +1527,10 @@ static void D_AddZip(const char* zipped_file_name, wad_source_t source)
 {
   dsda_string_t temporary_directory;
   char* full_zip_path;
+  static unsigned int file_counter = 0;
```

Comment:
Let's reorganize where the logic is and track things in the module. In order for that to make sense the module needs to do the decision making about where the zip file should be unzipped (i.e., the fact that it is a temporary directory). I would suggest the content of `D_AddZip` to be something like this:
```c
{
  char *temporary_directory;
  char *full_zip_path;

  full_zip_path = I_RequireZip(zipped_file_name);
  temporary_directory = dsda_UnzipFile(full_zip_path);

  if (!temporary_directory)
    I_Error("Unable to unzip %s", full_zip_path);

  LoadWADsAtPath(temporary_directory, source);
  LoadDehackedFilesAtPath(temporary_directory, true);

  Z_Free(temporary_directory);
}
```
The cleanup function can be called from the end of `I_EssentialQuit`

Reply:
Btw, just noticed this function is also missing `Z_Free(full_zip_path)` at the end.

Reply:
Thanks for the feedback. I've made the changed in [ece9a63](https://github.com/kraflab/dsda-doom/pull/353/commits/ece9a637a5b9310437d84d1271f948fd17d1c4e9).

Only thing I did not add was this:
```c
if (!temporary_directory)
    I_Error("Unable to unzip %s", full_zip_path);
```
Since `I_Error` would have been called already from the error that would make `dsda_UnzipFile` return `NULL`.

> Btw, just noticed this function is also missing `Z_Free(full_zip_path)` at the end.

Hadn't noticed the `I_Require*` needed to be freed oops 😅